File: utils/betManager.js
const { Collection } = require('discord.js');
const playerManager = require('./playerManager');
const raceManager = require('./raceManager');

// Thay đổi: bets lưu Array các vé cược thay vì object đơn
// Map<userId, Array<{horseNumber, amount, isAllIn}>>
let bets = new Collection();

async function getBalance(userId) {
    const player = await playerManager.getPlayer(userId);
    return player ? player.balance : 0;
}

// 1. Cập nhật hàm placeBet để xử lý nhiều vé cược
async function placeBet(userId, horseNumber, amount, isAllIn = false) {
  if (raceManager.isRaceInProgress()) {
    return { success: false, message: 'Không thể đặt cược khi cuộc đua đang diễn ra!' };
  }
  
  // Lấy danh sách các vé cược hiện tại của user (nếu chưa có thì tạo mảng mới)
  const userBets = bets.get(userId) || [];
  
  // KIỂM TRA: Giới hạn 3 vé cược
  if (userBets.length >= 3) {
    return { 
        success: false, 
        message: `Bạn chỉ được cược tối đa 3 mã ngựa! Bạn đã cược vào số: ${userBets.map(b => b.horseNumber).join(', ')}.` 
    };
  }
  
  if (horseNumber < 1 || horseNumber > raceManager.HORSE_COUNT) {
    return { success: false, message: `Số ngựa không hợp lệ (1-${raceManager.HORSE_COUNT}).` };
  }
  
  if (amount <= 0) {
    return { success: false, message: 'Tiền cược phải lớn hơn 0.' };
  }
  
  // Kiểm tra số dư
  if (!await playerManager.hasEnoughBalance(userId, amount)) {
    const player = await playerManager.getPlayer(userId);
    return { success: false, message: `Không có tiền mà cứ đòi cược lớn. ${player.balance} coin bọ :Đ.` };
  }
  
  // Trừ tiền ngay lập tức
  const newBalance = await playerManager.updateBalance(userId, -amount);
  
  // Lưu vé cược mới vào mảng
  userBets.push({
    horseNumber,
    amount,
    isAllIn
  });
  bets.set(userId, userBets);
  
  return {
    success: true,
    message: `Đã cược thêm **${amount} coin** vào ngựa **${horseNumber}**. (Vé ${userBets.length}/3)\nSố dư còn lại: ${newBalance}.`,
    balance: newBalance
  };
}

function getBet(userId) { return bets.get(userId) || null; }
function getAllBets() { return bets; }
function clearAllBets() { bets = new Collection(); }

// 2. Cập nhật hàm xử lý kết quả để duyệt qua từng vé cược
async function processBetResults(winners) {
  const results = [];
  
  // Duyệt qua từng người chơi
  for (const [userId, userBets] of bets.entries()) {
    // Duyệt qua từng vé cược của người chơi đó
    for (const bet of userBets) {
        const { horseNumber, amount, isAllIn } = bet;
        const isWinner = winners.includes(horseNumber);
        
        if (isWinner) {
          const winAmount = amount * 2;
          // Cộng tiền thắng
          const newBalance = await playerManager.updateBalance(userId, winAmount);
          
          results.push({
            userId, 
            won: true, 
            horseNumber, 
            betAmount: amount, 
            winAmount, 
            newBalance, 
            isAllIn
          });
        } else {
          // Thua thì không làm gì (tiền đã trừ lúc bet)
          const player = await playerManager.getPlayer(userId);
          results.push({
            userId, 
            won: false, 
            horseNumber, 
            betAmount: amount, 
            winAmount: 0, 
            newBalance: player.balance, 
            isAllIn
          });
        }
    }
  }
  return results;
}

// ... (Keep the code above unchanged)

// 2. Updated function to process bets for Rank 1 and Rank 2
async function processBetResults(rank1Horses, rank2Horses) {
  const results = [];
  
  // Iterate through each user
  for (const [userId, userBets] of bets.entries()) {
    // Iterate through each bet ticket of the user
    for (const bet of userBets) {
        const { horseNumber, amount, isAllIn } = bet;
        
        let winAmount = 0;
        let won = false;
        let rankType = null; // 'top1' or 'top2'

        // Check Top 1 (x2 Payout)
        if (rank1Horses.includes(horseNumber)) {
            won = true;
            winAmount = Math.floor(amount * 2);
            rankType = 'top1';
        } 
        // Check Top 2 (x1.5 Payout)
        else if (rank2Horses.includes(horseNumber)) {
            won = true;
            winAmount = Math.floor(amount * 1.5); // Floor to avoid decimals
            rankType = 'top2';
        }

        if (won) {
          // Add winning amount
          const newBalance = await playerManager.updateBalance(userId, winAmount);
          results.push({
            userId, 
            won: true, 
            horseNumber, 
            betAmount: amount, 
            winAmount, 
            newBalance, 
            isAllIn,
            rankType // Save rank type for display
          });
        } else {
          // Lost
          const player = await playerManager.getPlayer(userId);
          results.push({
            userId, 
            won: false, 
            horseNumber, 
            betAmount: amount, 
            winAmount: 0, 
            newBalance: player.balance, 
            isAllIn,
            rankType: null
          });
        }
    }
  }
  return results;
}

module.exports = { placeBet, getBet, getAllBets, clearAllBets, processBetResults, getBalance };

File: utils/bet_manager.py
import math
from typing import Optional
from utils import race_manager, player_manager
TOP1_MULTIPLIER: float = 3.6
TOP2_MULTIPLIER: float = 2.0
MAX_BETS_PER_USER: int = 3
_bets: dict[str, list[dict]] = {}

async def get_balance(user_id: str) -> int:
    player = await player_manager.get_player(user_id)
    return player['balance'] if player else 0

async def place_bet(user_id: str, horse_number: int, amount: int, is_all_in: bool=False) -> dict:
    if race_manager.is_race_in_progress():
        return {'success': False, 'message': 'Không thể đặt cược khi cuộc đua đang diễn ra!'}
    user_bets: list[dict] = _bets.get(user_id, [])
    if len(user_bets) >= MAX_BETS_PER_USER:
        return {'success': False, 'message': 'Bạn chỉ được cược tối đa 3 mã ngựa!'}
    if horse_number < 1 or horse_number > race_manager.HORSE_COUNT:
        return {'success': False, 'message': 'Số mã nương không hợp lệ.'}
    if amount <= 0:
        return {'success': False, 'message': 'Tiền cược phải lớn hơn 0.'}
    if not await player_manager.has_enough_balance(user_id, amount):
        player = await player_manager.get_player(user_id)
        balance = player['balance'] if player else 0
        return {'success': False, 'message': f'Không đủ tiền. Số dư: {balance}.'}
    new_balance: int = await player_manager.update_balance(user_id, -amount)
    user_bets.append({'horse_number': horse_number, 'amount': amount, 'is_all_in': is_all_in})
    _bets[user_id] = user_bets
    return {'success': True, 'message': f'Đã ủng hộ thêm **{amount} coin** cho mã nương **{horse_number}**. (Vé {len(user_bets)}/{MAX_BETS_PER_USER})\nSố dư còn lại: {new_balance}.', 'balance': new_balance}

def get_bet(user_id: str) -> Optional[list[dict]]:
    return _bets.get(user_id) or None

def get_all_bets() -> dict[str, list[dict]]:
    return _bets

def clear_all_bets() -> None:
    _bets.clear()

async def process_bet_results(rank1_horses: list[int], rank2_horses: list[int]) -> list[dict]:
    results: list[dict] = []
    for user_id, user_bets in _bets.items():
        for bet in user_bets:
            horse_number: int = bet['horse_number']
            amount: int = bet['amount']
            is_all_in: bool = bet['is_all_in']
            win_amount: int = 0
            won: bool = False
            rank_type: Optional[str] = None
            if horse_number in rank1_horses:
                won = True
                win_amount = math.floor(amount * TOP1_MULTIPLIER)
                rank_type = 'top1'
            elif horse_number in rank2_horses:
                won = True
                win_amount = math.floor(amount * TOP2_MULTIPLIER)
                rank_type = 'top2'
            if won:
                new_balance = await player_manager.update_balance(user_id, win_amount)
                results.append({'userId': user_id, 'won': True, 'horseNumber': horse_number, 'betAmount': amount, 'winAmount': win_amount, 'newBalance': new_balance, 'isAllIn': is_all_in, 'rankType': rank_type})
            else:
                player = await player_manager.get_player(user_id)
                current_balance = player['balance'] if player else 0
                results.append({'userId': user_id, 'won': False, 'horseNumber': horse_number, 'betAmount': amount, 'winAmount': 0, 'newBalance': current_balance, 'isAllIn': is_all_in, 'rankType': None})
    return results